Fails to build with GCC 12 (all versions)

## Issue Description

Fails to build stellar-core with GCC 12

## Steps to Reproduce

Run build on Fedora 36

## Expected Result
build success

## Actual Result

```
g++ -std=c++17 -DHAVE_CONFIG_H -I. -I..  -isystem ".." -I"../src" -I"../src" -isystem /builddir/build/BUILD/stellar-core-18.4.0/lib/libsodium/src/libsodium/include -isystem ../lib/libsodium/src/libsodium/include -isystem ../lib/libsodium/src/libsodium/include/sodium -isystem /builddir/build/BUILD/stellar-core-18.4.0/lib/xdrpp -isystem /builddir/build/BUILD/stellar-core-18.4.0/lib/xdrpp -isystem ../lib/libmedida/src -isystem ../lib/soci/src/core -isystem ../lib/sqlite -DSQLITE_CORE -DSQLITE_OMIT_LOAD_EXTENSION=1 -DASIO_SEPARATE_COMPILATION=1 -DASIO_STANDALONE -isystem ../lib/asio/asio/include  -isystem "../lib" -isystem "../lib/autocheck/include" -isystem "../lib/cereal/include" -isystem "../lib/util" -isystem "../lib/fmt/include" -isystem "../lib/soci/src/core" -isystem "../lib/tracy" -isystem "../lib/spdlog/include" -DUSE_POSTGRES=1   -DBUILD_TESTS=1 -DUSE_SPDLOG    -O2 -flto=auto -ffat-lto-objects -fexceptions -g -grecord-gcc-switches -pipe -Wall -Werror=format-security -Wp,-D_FORTIFY_SOURCE=2 -Wp,-D_GLIBCXX_ASSERTIONS -specs=/usr/lib/rpm/redhat/redhat-hardened-cc1 -fstack-protector-strong -specs=/usr/lib/rpm/redhat/redhat-annobin-cc1  -m64  -mtune=generic -fasynchronous-unwind-tables -fstack-clash-protection -fcf-protection -pthread -DFMT_HEADER_ONLY=1 -Wno-unused-command-line-argument -Wno-unused-local-typedef -Wno-unknown-warning-option -Werror=unused-result -c -o util/Scheduler.o util/Scheduler.cpp
In file included from util/Scheduler.cpp:6:
../lib/util/finally.h: In constructor 'gsl::final_action<F>::final_action(gsl::final_action<F>&&)':
../lib/util/finally.h:37:49: error: 'exchange' is not a member of 'std'
   37 |         : f_(std::move(other.f_)), invoke_(std::exchange(other.invoke_, false))
      |                                                 ^~~~~~~~
At global scope:
cc1plus: note: unrecognized command-line option '-Wno-unknown-warning-option' may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option '-Wno-unused-local-typedef' may have been intended to silence earlier diagnostics
cc1plus: note: unrecognized command-line option '-Wno-unused-command-line-argument' may have been intended to silence earlier diagnostics
make[3]: *** [Makefile:3094: util/Scheduler.o] Error 1
```

## Your Environment and Setup

```
Installing:
 automake                    noarch   1.16.5-4.fc36              fedora   673 k
 bison                       x86_64   3.8.2-2.fc36               fedora   986 k
 clang                       x86_64   13.0.1-1.fc36              fedora    84 k
 flex                        x86_64   2.6.4-10.fc36              fedora   307 k
 gcc-c++                     x86_64   12.0.1-0.12.fc36           fedora    13 M
 git                         x86_64   2.35.1-1.fc36              fedora    67 k
 hostname                    x86_64   3.23-6.fc36                fedora    27 k
 libpq-devel                 x86_64   14.1-2.fc36                fedora    96 k
 libtool                     x86_64   2.4.6-50.fc36              fedora   579 k
 libunwind-devel             x86_64   1.6.2-2.fc36               fedora    81 k
 parallel                    noarch   20211222-3.fc36            fedora   387 k
 postgresql-server           x86_64   14.1-3.fc36                fedora   6.0 M
 systemd-rpm-macros          noarch   250.3-8.fc36               fedora    29 k
Installing dependencies:
 annobin-docs                noarch   10.57-3.fc36               fedora    89 k
 annobin-plugin-gcc          x86_64   10.57-3.fc36               fedora   877 k
 autoconf                    noarch   2.71-2.fc36                fedora   718 k
 clang-libs                  x86_64   13.0.1-1.fc36              fedora    23 M
 clang-resource-filesystem   x86_64   13.0.1-1.fc36              fedora    14 k
 cpp                         x86_64   12.0.1-0.12.fc36           fedora    10 M
 dbus                        x86_64   1:1.14.0-1.fc36            fedora   7.5 k
 dbus-broker                 x86_64   29-5.fc36                  fedora   174 k
 dbus-common                 noarch   1:1.14.0-1.fc36            fedora    14 k
 emacs-filesystem            noarch   1:27.2-9.fc35              fedora   9.0 k
 expat                       x86_64   2.4.7-1.fc36               fedora   109 k
 gc                          x86_64   8.0.6-2.fc36               fedora   103 k
 gcc                         x86_64   12.0.1-0.12.fc36           fedora    33 M
 git-core                    x86_64   2.35.1-1.fc36              fedora   3.9 M
 git-core-doc                noarch   2.35.1-1.fc36              fedora   2.4 M
 glibc-devel                 x86_64   2.35-4.fc36                fedora   154 k
 glibc-headers-x86           noarch   2.35-4.fc36                fedora   557 k
 groff-base                  x86_64   1.22.4-9.fc36              fedora   1.0 M
 guile22                     x86_64   2.2.7-5.fc36               fedora   6.4 M
 kernel-headers              x86_64   5.17.0-300.fc36            fedora   1.3 M
 kmod-libs                   x86_64   29-7.fc36                  fedora    68 k
 less                        x86_64   590-3.fc36                 fedora   160 k
 libcbor                     x86_64   0.7.0-5.fc36               fedora    56 k
 libedit                     x86_64   3.1-41.20210910cvs.fc36    fedora   104 k
 libfido2                    x86_64   1.10.0-3.fc36              fedora    95 k
 libicu                      x86_64   69.1-5.fc36                fedora   9.8 M
 libmpc                      x86_64   1.2.1-4.fc36               fedora    61 k
 libpq                       x86_64   14.1-2.fc36                fedora   198 k
 libseccomp                  x86_64   2.5.3-2.fc36               fedora    70 k
 libstdc++-devel             x86_64   12.0.1-0.12.fc36           fedora   2.3 M
 libtool-ltdl                x86_64   2.4.6-50.fc36              fedora    36 k
 libunwind                   x86_64   1.6.2-2.fc36               fedora    67 k
 libxcrypt-devel             x86_64   4.4.28-1.fc36              fedora    29 k
 llvm                        x86_64   13.0.1-1.fc36              fedora    13 M
 llvm-libs                   x86_64   13.0.1-1.fc36              fedora    23 M
```

## stellar-core Version

all versions, file https://github.com/stellar/stellar-core/blob/v18.5.0/lib/util/finally.h#L37 has not changed for the last 2 years


## Additional context

i think the issue related with this changes 

http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2021/p2401r0.html

https://github.com/cplusplus/draft/pull/4999

https://reviews.llvm.org/D111481

how to reproduce:

```Dockerfile
# file issue3388.Containerfile
FROM docker.io/library/fedora:36
RUN dnf install gcc-g++ git -q -y
RUN git clone --quiet --depth 1 --recurse-submodules --shallow-submodules https://github.com/stellar/stellar-core.git
WORKDIR /stellar-core
RUN g++  -I"src/" -I. -I"lib/asio/asio/include/" -I"lib/tracy/"  src/util/Scheduler.cpp
```

```shell
$ podman build -f issue3388.Containerfile
STEP 1/5: FROM docker.io/library/fedora:36
STEP 2/5: RUN dnf install gcc-g++ git -q -y
--> 8aca35823ea
STEP 3/5: RUN git clone --quiet --depth 1 --recurse-submodules --shallow-submodules https://github.com/stellar/stellar-core.git
--> 3e45d2ce21c
STEP 4/5: WORKDIR /stellar-core
--> 929643377a1
STEP 5/5: RUN g++  -I"src/" -I. -I"lib/asio/asio/include/" -I"lib/tracy/"  src/util/Scheduler.cpp
In file included from src/util/Scheduler.cpp:6:
./lib/util/finally.h: In constructor 'gsl::final_action<F>::final_action(gsl::final_action<F>&&)':
./lib/util/finally.h:37:49: error: 'exchange' is not a member of 'std'
   37 |         : f_(std::move(other.f_)), invoke_(std::exchange(other.invoke_, false))
      |                                                 ^~~~~~~~
Error: error building at STEP "RUN g++  -I"src/" -I. -I"lib/asio/asio/include/" -I"lib/tracy/"  src/util/Scheduler.cpp": error while running runtime: exit status 1
$
```

[gcc-12.1](https://gcc.gnu.org/gcc-12/) released six days ago.
how to reproduce with GCC official docker image:


```shell
$ docker run -it --rm  docker.io/library/gcc:12.1.0

root@fa4c9d63ad29:/# git clone -q  --depth 1 --recurse-submodules --shallow-submodules https://github.com/stellar/stellar-core.git 2>/dev/null
root@fa4c9d63ad29:/# cd stellar-core/ && git log -n1
commit ce99e88f0e3615d83b75f80286b3000a77afe27c (grafted, HEAD -> master, origin/master, origin/HEAD)
Author: Latobarita <latobarita@users.noreply.github.com>
Date:   Wed May 11 12:15:31 2022 -0700
    Merge pull request #3421 from dmkozh/patch-1
    Clarified the required Rust toolchain in Win docs
    Reviewed-by: MonsieurNicolas


root@fa4c9d63ad29:/stellar-core# g++  -I"src/" -I. -I"lib/asio/asio/include/" -I"lib/tracy/"  src/util/Scheduler.cpp
In file included from src/util/Scheduler.cpp:6:
./lib/util/finally.h: In constructor 'gsl::final_action<F>::final_action(gsl::final_action<F>&&)':
./lib/util/finally.h:37:49: error: 'exchange' is not a member of 'std'
   37 |         : f_(std::move(other.f_)), invoke_(std::exchange(other.invoke_, false))
      |                                                 ^~~~~~~~
root@fa4c9d63ad29:/stellar-core# 
```

--- a/lib/util/finally.h
+++ b/lib/util/finally.h
@@ -22,6 +22,7 @@
 // We therefore keep it in the same 'gsl' namespace it appeared in in the
 // original file.
 
+#include <utility>
 namespace gsl
 {
 

